### PR TITLE
Generalize span creation 

### PR DESCRIPTION
**Pull request by c24t**
Updates #165.

This PR removes references to the java-specific `SpanBuilder` in the span creation spec, and borrows from the java documentation and [OpenTelemetry spec](https://opentracing.io/specification/).

@songy23, @bogdandrutu I'd like to get your feedback here before moving on to span operations.

### CODE REVIEW

File: specification/tracing-api.md [hunk 1]
```diff
@@ -200,82 +200,52 @@ Please review the W3C specification for details on the [Tracestate field](https:
...
-- Name of the span.
+### Span Creation
+
+Implementations MUST provide a way to create `Span`s via a `Tracer`, which is responsible for tracking the currently active `Span` and MAY provide default options for newly created `Span`s.
+
+The `Tracer` SHOULD create each new `Span` as a child of its active `Span`, and should set the newly created `Span` as active.
+The `Tracer` MUST allow the caller to specify the new `Span`'s parent in the form of a `Span` or `SpanContext`.
+If the caller does explicitly set the `Span`'s parent, the `Tracer`'s currently active `Span` SHOULD NOT change.
+
+Each span has zero or one parent spans and zero or more child spans, which represent causally related operations.
```
Comment (reyang) [inline specification/tracing-api.md]: "one parent span" (removing `s`)

File: specification/tracing-api.md [hunk 2]
```diff
@@ -200,82 +200,52 @@ Please review the W3C specification for details on the [Tracestate field](https:
 
 ## Span
 
-Span represents a single operation within a trace. Spans can be nested to form a
-trace tree. Often, a trace contains a root span that describes the end-to-end
-latency and, optionally, one or more sub-spans for its sub-operations.
+A `Span` represents a single operation within a trace.
+Spans can be nested to form a trace tree.
+Each trace contains a root span, which typically describes the end-to-end latency and, optionally, one or more sub-spans for its sub-operations.
```
Comment (bogdandrutu) [inline specification/tracing-api.md]: we should use a limit for every line. I think the current vscode config asks for 80.

File: specification/tracing-api.md [hunk 3]
```diff
@@ -200,82 +200,52 @@ Please review the W3C specification for details on the [Tracestate field](https:
...
+Vendors may implement the `Span` interface to effect vendor-specific logic.
+However, alternative implementations MUST NOT allow callers to create `Span`s directly.
+All `Span`s MUST be created via a `Tracer`.
 
-- Name of the span.
+### Span Creation
+
+Implementations MUST provide a way to create `Span`s via a `Tracer`, which is responsible for tracking the currently active `Span` and MAY provide default options for newly created `Span`s.
+
+The `Tracer` SHOULD create each new `Span` as a child of its active `Span`, and should set the newly created `Span` as active.
```
Comment (bogdandrutu) [inline specification/tracing-api.md]: We do the opposite in Java. We do not set as `active` if by active means in the current context. we do want the user to explicitly call `withSpan`, because we do need to support the case when the new span is not set as active.

Comment (bogdandrutu) [inline specification/tracing-api.md]: I would say:
The `Tracer` SHOULD create each new `Span` as a child of its active `Span` unless an explicit parent is provided.

File: specification/tracing-api.md [hunk 4]
```diff
@@ -200,82 +200,52 @@ Please review the W3C specification for details on the [Tracestate field](https:
...
-follow its own convention on `Span` creation, for example `Builder` in Java,
-`Options` in Go, etc. `Span` creation method MUST be defined on `Tracer`.
 
-Required parameters:
+Vendors may implement the `Span` interface to effect vendor-specific logic.
+However, alternative implementations MUST NOT allow callers to create `Span`s directly.
+All `Span`s MUST be created via a `Tracer`.
 
-- Name of the span.
+### Span Creation
```
Comment (bogdandrutu) [inline specification/tracing-api.md]: I would summarize all the configs that we want users to be able to provide, currently it is hard to find this list because every option is introduced in one paragraph.

```markdown
### Span Creation
Implementations MUST provide a way to create `Span`s via a `Tracer`, which is responsible for tracking the currently active `Span` and MAY provide default options for newly created `Span`s.

The users of this API SHOULD be able to provide the following properties:
* A parent Span (using a Span, SpanContext, or no parent).
* Parent links.
* ...
```

File: specification/tracing-api.md [hunk 5]
```diff
@@ -200,82 +200,52 @@ Please review the W3C specification for details on the [Tracestate field](https:
...
+`Span`s encapsulate:
 
-The only two getters on span returns `SpanContext` and the flag on whether span
-will be recorded.
+- The operation name
+- A [`SpanContext`](#SpanContext)
+- A parent [`Span`](#Span) or [`SpanContext`](#SpanContext)
+- A start timestamp
+- An end timestamp
+- A set of [`Attribute`s](#SetAttribute)
```
Comment (songy23) [inline specification/tracing-api.md]: Do we want to keep the order when `Attribute`/`Link`/`Event` is added? (In Java we do.) If so this should be "A link of `Attribute`/`Link`/`Event`".

File: specification/tracing-api.md (hunk 2 above)
Comment (c24t) [reply specification/tracing-api.md]: Changed in a036be3 and ed16890.

We're not doing this consistently across the repo, e.g. https://github.com/open-telemetry/opentelemetry-specification/pull/186 uses a single line for each paragraph.

I think there's a good argument for soft-wrapping sentences or paragraphs, especially since putting each sentence on its own line means most edits show up as line diffs instead of spilling over into other lines.

I'm pretty surprised to see a vscode config file in this repo.

Comment (c24t) [reply specification/tracing-api.md]: [third_way](https://xkcd.com/1285/)

File: specification/tracing-api.md (hunk 3 above)
Comment (c24t) [reply specification/tracing-api.md]: Changed in 7e9984f.

Comment (c24t) [reply specification/tracing-api.md]: Changed in 7e9984f.

File: specification/tracing-api.md (hunk 5 above)
Comment (c24t) [reply specification/tracing-api.md]: Good call, fixed in 2742bc7.

I called attributes "an ordered mapping". This may be too prescriptive, but "list" is misleading since the whole point of attributes is that you can fetch them by key.

File: specification/tracing-api.md (hunk 4 above)
Comment (c24t) [reply specification/tracing-api.md]: Changed in a4a8e38 and reworded to make it less redundant with the list above, let me know if this is what you have in mind.

File: specification/tracing-api.md [hunk 6]
```diff
@@ -89,29 +89,35 @@ A duration is the elapsed time between two events.
...
+provided externally.
 
 #### Tracer provider
 
-Tracer provider is an internal class used by the global registry (`OpenTelemetry`) to get a tracer instance.
-The global registry delegates calls to the provider every time a tracer instance is requested.
-This is necessary for use-cases when a single instrumentation code runs for multiple deployments.
+Tracer provider is an internal class used by the global registry
+(`OpenTelemetry`) to get a tracer instance. The global registry delegates calls
+> to the provider every time a tracer instance is requested. This is necessary
```
Comment (arminru) [inline specification/tracing-api.md]:
```suggestion
to the provider every time a tracer instance is requested. This is necessary
```
unintended block quote

File: specification/tracing-api.md [hunk 7]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
 
-Span represents a single operation within a trace. Spans can be nested to form a
-trace tree. Often, a trace contains a root span that describes the end-to-end
-latency and, optionally, one or more sub-spans for its sub-operations.
+A `Span` represents a single operation within a trace. Spans can be nested to
+form a trace tree. Each trace contains a root span, which typically describes
+the end-to-end latency and, optionally, one or more sub-spans for its
+sub-operations.
+
+`Span`s encapsulate:
```
Comment (arminru) [inline specification/tracing-api.md]: great overview! 👍

File: specification/tracing-api.md [hunk 8]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
+The API MUST allow users to provide the following properties:
+- The operation name
+- The parent span, and whether the new `Span` should be a root `Span`.
+- `Attribute`s
+- `Link`s
+- `Event`s
+
+The `Tracer` MUST allow the caller to specify the new `Span`'s parent in the
+form of a `Span` or `SpanContext`. The `Tracer` SHOULD create each new `Span` as
+a child of its active `Span` unless an explicit parent is provided.
```
Comment (arminru) [inline specification/tracing-api.md]:
```suggestion
a child of its active `Span` unless an explicit parent is provided or the
option to create a span without a parent is selected.```

File: specification/tracing-api.md [hunk 9]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
+
+Each span has zero or one parent span and zero or more child spans, which
+represent causally related operations. A tree of related spans comprises a
+trace. A span is said to be a _root span_ if it does not have a parent. Each
+trace includes a single root span, which is the shared ancestor of all other
+spans in the trace. Implementations MUST provide an option to create a `Span` as
+a root span, and MUST generate a new `TraceId` for each root span created.
+
+A `Span` is said to have a _remote parent_ if it is the child of a `Span`
+created in another process.Since the `SpanContext` is the only component of a
```
Comment (arminru) [inline specification/tracing-api.md]:
```suggestion
created in another process. Since the `SpanContext` is the only component of a
```

File: specification/tracing-api.md [hunk 10]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
+`Span` that is propagated between processes, a `Span`'s parent SHOULD be a
+`SpanContext` if it is remote.Otherwise, it may be a `Span` or `SpanContext`.
```
Comment (arminru) [inline specification/tracing-api.md]:
```suggestion
`SpanContext` if it is remote. Otherwise, it may be a `Span` or `SpanContext`.
```

File: specification/tracing-api.md [hunk 11]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
+Implementations MUST provide a way to create `Span`s via a `Tracer`, which is
+responsible for tracking the currently active `Span` and MAY provide default
+options for newly created `Span`s.
+
+The API MUST allow users to provide the following properties:
+- The operation name
+- The parent span, and whether the new `Span` should be a root `Span`.
+- `Attribute`s
+- `Link`s
+- `Event`s
```
Comment (bogdandrutu) [inline specification/tracing-api.md]: Is this new?

File: specification/tracing-api.md [hunk 12]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
+directly. All `Span`s MUST be created via a `Tracer`.
+
+### Span Creation
+
+Implementations MUST provide a way to create `Span`s via a `Tracer`, which is
+responsible for tracking the currently active `Span` and MAY provide default
+options for newly created `Span`s.
+
+The API MUST allow users to provide the following properties:
+- The operation name
```
Comment (bogdandrutu) [inline specification/tracing-api.md]: This should be marked as required, the others are optional (specify the default).

File: specification/tracing-api.md [hunk 13]
```diff
@@ -167,115 +176,120 @@ Returns the binary format interface which can serialize/deserialize `Span`s.
...
+`Span`s encapsulate:
+
+- The operation name
+- An immutable [`SpanContext`](#SpanContext) that uniquely identifies the
+  `Span`
+- A parent span in the form of a [`Span`](#Span), [`SpanContext`](#SpanContext),
+  or null
+- A start timestamp
+- An end timestamp
+- An ordered mapping of [`Attribute`s](#SetAttribute)
```
Comment (reyang) [inline specification/tracing-api.md]: Do we want to enforce ordering in the spec?

I understand that the actual implementation might put a limit on the maximum number of attributes, and in order to achieve this, the implementation could choose ordered mapping to implement FIFO. But is this a scenario/semantic we want to mention in the spec or it is just implementation details?

Comment (songy23) [reply specification/tracing-api.md]: Ordering won't be reflected directly in the APIs but it still affects the overall behavior. I think it's worth mentioning in the spec so that all implementation can conform to it.

Comment (reyang) [reply specification/tracing-api.md]: Let's say, if there is an implementation which supports unlimited number of attributes, do we still want it to maintain the order? What would the consumer do with the ordering?

If we do have ordering requirement, it might be worthy to mention `implementation which puts limit on the number of key-value-pairs should follow FIFO`.
Another detail that we might want to cover, if there is a duplicated key but different value, what's the intended behavior. For links/events, if the user is trying to append a value which already exists, should the library de-dupe, report error, or append?

Comment (songy23) [reply specification/tracing-api.md]: >  if there is an implementation which supports unlimited number of attributes, do we still want it to maintain the order? 

> If we do have ordering requirement, it might be worthy to mention implementation which puts limit on the number of key-value-pairs should follow FIFO.

Agreed if there's no limit specified for attributes, rephrase it to FIFO may be clearer. However I think preserving the order for timed events makes more sense. To simplify things I would recommend we keep the same for events, links and attributes.

> For links/events, if the user is trying to append a value which already exists, should the library de-dupe, report error, or append?

I would say the library would just append instead of de-dup. For events specifically, it's difficult to determine if two events are duplicated since events depend on the time when they are added. 

Would like to hear others' opinions on this.

File: specification/tracing-api.md (hunk 11 above)
Comment (c24t) [reply specification/tracing-api.md]: It's the items from the list above that are set by the user. I added this in response to https://github.com/open-telemetry/opentelemetry-specification/pull/190#discussion_r306242860.

File: specification/tracing-api.md (hunk 12 above)
Comment (c24t) [reply specification/tracing-api.md]: How about this?

```markdown
The API SHOULD require the caller to provide:
- The operation name
- The parent span, and whether the new `Span` should be a root `Span`.

The API MUST allow users to provide the following properties, which SHOULD be
empty by default:
- `Attribute`s
- `Link`s
- `Event`s
```

File: specification/tracing-api.md (hunk 13 above)
Comment (c24t) [reply specification/tracing-api.md]: I think it makes sense to preserve ordering, especially since it may be exporters instead of the span impl that's responsible for truncating these. In that case we'd use the ordering even if there's no limit in the span.

> I would say the library would just append instead of de-dup

I'd expect attributes tobe de-duped since (this version of) the spec says they're stored in a map, events and links get appended to the list.

Also "ordering" here doesn't specify anything _about_ the order, just that it's deterministic.

File: specification/tracing-api.md (hunk 10 above)
Comment (c24t) [reply specification/tracing-api.md]: Thanks for catching these!

File: specification/tracing-api.md (hunk 13 above)
Comment (songy23) [reply specification/tracing-api.md]: Ah that's correct, for attributes we should do de-dup.

Comment (reyang) [reply specification/tracing-api.md]: Regarding de-dupe on the order mapping, we also need to mention the ordering (the latest value of the same key would win, and the key-value pair will be treated as the newest inserted one). We could borrow idea from [TraceContext spec](https://github.com/w3c/trace-context/blob/master/spec/20-http_header_format.md#combined-header-value).

Comment (c24t) [reply specification/tracing-api.md]: I added a note on de-duping under "Set Attributes" in d2cebe9. I'll edit the span operations section next and expand on this there.

**Comment by c24t** [comment]
This should be ready to merge.

**Comment by carlosalberto** [comment]
Merging now, as this has been approved for a few days now. We can always receive feedback later on ;)
